runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt

```diff
@@ -0,0 +1,205 @@
+
+        val serializedWithNull = DynamicObjectSerializer(encodeDefaults = false).serialize(data)
+        assertNull(serializedWithNull.d)
+        assertTrue(js("""Object.keys(serializedWithNull).includes("d")"""), "should contain null properties")
+        Json(JsonConfiguration())
+    }
+
+    @Test
+    fun listTest() {
+        assertDynamicForm(listOf(1, 2, 3, 44), serializer = IntSerializer.list) { data, serialized ->
+            assertNotNull(serialized.length, "length property should exist")
+            assertEquals(data.size, serialized.length)
+
+            for (i in data.indices) {
+                assertEquals(data[i], serialized[i])
+            }
+        }
+    }
+
+    @Test
+    fun arrayTest() {
+        assertDynamicForm(intArrayOf(1, 2, 3, 44), serializer = IntArraySerializer) { data, serialized ->
+            assertNotNull(serialized.length, "length property should exist")
+            assertEquals(data.size, serialized.length)
+
+            for (i in data.indices) {
+                assertEquals(data[i], serialized[i])
+            }
+        }
+    }
+
+    @Test
+    fun nestedListTest() {
+        assertDynamicForm(NestedList("a string", listOf(1, 2, 3, 44))) { data, serialized ->
+            assertEquals(data.a, serialized.a)
+            assertNotNull(serialized.list.length, "length property should exist")
+            assertEquals(data.list.size, serialized.list.length)
+
+            for (i in data.list.indices) {
+                assertEquals(data.list[i], serialized.list[i])
+            }
+        }
+
+    }
+
+    @Test
+    fun complexMapWrapperTest() {
+        assertDynamicForm(ComplexMapWrapper(mapOf("key1" to Data(1), "key2" to Data(2))))
+    }
+
+    @Test
+    fun mapWrapperTest() {
+        assertDynamicForm(MapWrapper(mapOf("key1" to 1, "key2" to 2)))
+    }
+
+    @Test
+    fun listOfListsTest() {
+        assertDynamicForm(
+            ListOfLists(
+                listOf(
+                    listOf(Data(11), Data(12), Data(13)),
+                    listOf(Data(21), Data(22))
+                )
+            )
+        ) { data, serialized ->
+            assertEquals(data.l.size, serialized.l.length)
+            assertEquals(data.l.first().size, serialized.l[0].length)
+        }
+    }
+
+    @Test
+    fun nestedCollections() {
+        @Serializable
+        class NestedCollections(val data: Map<String, Map<String, List<Int>>>)
+
+        assertDynamicForm(
+            NestedCollections(
+                mapOf(
+                    "one" to mapOf("oneone" to listOf(11, 12, 13), "onetwo" to listOf(1)),
+                    "two" to mapOf("twotwo" to listOf(22, 23))
+                )
+            )
+            , serializer = NestedCollections.serializer()
+        )
+    }
+
+    @Test
+    fun enums() {
+        assertDynamicForm(EnumWrapper(Color.RED))
+        assertDynamicForm(Color.GREEN, serializer = CommonEnumSerializer("Color"))
```

Comment:
The default `CommonEnumSerializer` seems to just serialize using `name`, so I don't think this test is a good indicator for using custom enum serializers.

Can we replace this test with a serializer that behaves differently from the default behavior?

Maybe using something like 

```kt
val serializer = CommonEnumSerializer(
  serialName, 
  enumValues(), 
  enumValues<Color>()
    .map { "${it.name}_${it.ordinal}" }
    .toTypedArray()
)
```

Reply:
The test case may have been confusing, I only wanted to test the serialization of a singular enum. I will add `@Serializable` to the Color enum and remove the custom one here.

We can also add an additional test with custom serialisers.